### PR TITLE
HTTP API Implementation for Opportunities & Suggestions

### PR DESCRIPTION
**Pull request by chicharr**
HTTP API Implementation for Opportunities & Suggestions

Solves: https://jira.corp.adobe.com/browse/SITES-26860

Please ensure your pull request adheres to the following guidelines:
- [X] make sure to link the related issues in this description. Or if there's no issue created, make sure you 
  describe here the problem you're solving.
- [X] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

If the PR is changing the API specification:
- [ ] make sure you add a "Not implemented yet" note the endpoint description, if the implementation is not ready 
  yet. Ideally, return a 501 status code with a message explaining the feature is not implemented yet.
- [ ] make sure you add at least one example of the request and response.

If the PR is changing the API implementation or an entity exposed through the API:
- [X] make sure you update the API specification and the examples to reflect the changes. -> in a separate PR

If the PR is introducing a new audit type:
- [ ] make sure you update the API specification with the type, schema of the audit result and an example

## Related Issues

Thanks for contributing!

### CODE REVIEW

File: src/controllers/opportunities.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+ * @returns {object} Opportunities controller.
+ * @constructor
+ */
+function OpportunitiesController(dataAccess) {
+  if (!isObject(dataAccess)) {
+    throw new Error('Data access required');
+  }
+  const { Opportunity } = dataAccess;
+  if (!isObject(Opportunity)) {
+    throw new Error('Data access required');
```
Comment (solaris007) [inline src/controllers/opportunities.js:41]: should be different error message

File: src/controllers/opportunities.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  };
+
+  /**
+   * Gets all opportunities for a given site type filtering by status.
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} Array of opportunities response.
+   */
+  const getByStatus = async (context) => {
+    const siteId = context.params?.siteId;
+    const status = context.params?.status || undefined;
```
Comment (solaris007) [inline src/controllers/opportunities.js:69]: why is this branch needed?

File: src/controllers/opportunities.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  };
+
+  /**
+   * Gets an opportunity for a given site type and opportunity ID.
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} Opportunity response.
+   */
+  const getByID = async (context) => {
+    const siteId = context.params?.siteId;
+    const opptyId = context.params?.opportunityId || undefined;
```
Comment (solaris007) [inline src/controllers/opportunities.js:91]: why is this branch needed?

File: src/controllers/opportunities.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    }
+    if (!isNonEmptyObject(context.data)) {
+      return badRequest('No data provided');
+    }
+
+    context.data.siteId = siteId;
+    try {
+      const oppty = await Opportunity.create(context.data);
+      return createResponse(OpportunityDto.toJSON(oppty), 201);
+    } catch (e) {
```
Comment (solaris007) [inline src/controllers/opportunities.js:126]: bad request would only be applicable if a validation error occurs, but there might also be other errors

File: src/controllers/opportunities.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+      if (tags && !arrayEquals(tags, opportunity.getTags())) {
+        hasUpdates = true;
+        opportunity.setTags(tags);
+      }
+      if (hasUpdates) {
+        const updatedOppty = await opportunity.save(opportunity);
+        return ok(OpportunityDto.toJSON(updatedOppty));
+      }
+    } catch (e) {
+      return badRequest(e.message);
```
Comment (solaris007) [inline src/controllers/opportunities.js:199]: bad request would only be applicable if a validation error occurs, but there might also be other errors

File: src/controllers/opportunities.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    // TODO: eventually suggestions will be removed by the data access layer
+    // and this code should be removed
+    const arraySuggestions = await opportunity.getSuggestions();
+    const removeSuggestionsPromises = arraySuggestions
+      .map(async (suggestion) => suggestion.remove());
+
+    await Promise.all(removeSuggestionsPromises);
+
+    await opportunity.remove();
+    return noContent();
```
Comment (solaris007) [inline src/controllers/opportunities.js:235]: missing error handling

File: src/controllers/suggestions.js [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    if (!hasText(opptyId)) {
+      return badRequest('Opportunity ID required');
+    }
+
+    const suggestionEntities = await Suggestion.allByOpportunityId(opptyId);
+    // Check if the opportunity belongs to the site
+    if (suggestionEntities.length > 0) {
+      const oppty = await suggestionEntities[0].getOpportunity();
+      if (!oppty || oppty.getSiteId() !== siteId) {
+        return badRequest('Opportunity not found');
```
Comment (solaris007) [inline src/controllers/suggestions.js:64]: shouldn't this be a `notFound` instead?

File: src/controllers/suggestions.js [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    }
+
+    if (!hasText(opptyId)) {
+      return badRequest('Opportunity ID required');
+    }
+
+    const suggestionEntities = await Suggestion.allByOpportunityId(opptyId);
+    // Check if the opportunity belongs to the site
+    if (suggestionEntities.length > 0) {
+      const oppty = await suggestionEntities[0].getOpportunity();
```
Comment (solaris007) [inline src/controllers/suggestions.js:62]: i suggest getting the oppty earlier and adding a 404 guard statement

File: src/controllers/suggestions.js [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+      return badRequest('Opportunity ID required');
+    }
+    if (!hasText(status)) {
+      return badRequest('Status is required');
+    }
+
+    const suggestionEntities = await Suggestion.allByOpportunityIdAndStatus(opptyId, status);
+    // Check if the opportunity belongs to the site
+    if (suggestionEntities.length > 0) {
+      const oppty = await suggestionEntities[0].getOpportunity();
```
Comment (solaris007) [inline src/controllers/suggestions.js:93]: see above

File: src/controllers/suggestions.js [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+      return badRequest('Suggestion ID required');
+    }
+
+    const suggestion = await Suggestion.findById(suggestionId);
+    if (!suggestion || suggestion.getOpportunityId() !== opptyId) {
+      return notFound('Suggestion not found');
+    }
+    const opportunity = await suggestion.getOpportunity();
+    if (!opportunity || opportunity.getSiteId() !== siteId) {
+      return notFound('Suggestion not found');
```
Comment (solaris007) [inline src/controllers/suggestions.js:130]: Opportunity not found?

File: src/controllers/suggestions.js [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+    // validate request body
+    if (!context.data) {
+      return badRequest('No updates provided');
+    }
+
+    if (!isArray(context.data)) {
+      return badRequest('Request body must be an array');
+    }
+
+    const suggestionPromises = context.data.map(async (suggData, index) => {
```
Comment (solaris007) [inline src/controllers/suggestions.js:161]: use `Suggestion.createMany` instead ...

File: src/controllers/suggestions.js [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+        hasUpdates = true;
+        suggestion.setKpiDeltas(kpiDeltas);
+      }
+
+      if (hasUpdates) {
+        const updatedSuggestion = await suggestion.save();
+        return ok(SuggestionDto.toJSON(updatedSuggestion));
+      }
+    } catch (e) {
+      return badRequest(e.message);
```
Comment (solaris007) [inline src/controllers/suggestions.js:255]: bad request would only be applicable if a validation error occurs, but there might also be other errors

File: src/controllers/suggestions.js [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+    // validate request body
+    if (!context.data) {
+      return badRequest('No updates provided');
+    }
+
+    if (!isArray(context.data)) {
+      return badRequest('Request body must be an array of [{ id: <suggestion id>, status: <suggestion status> },...]');
+    }
+
+    const suggestionPromises = context.data.map(async ({ id, status }, index) => {
```
Comment (solaris007) [inline src/controllers/suggestions.js:286]: use `Suggestion.bulkUpdateStatus` instead?

File: src/controllers/suggestions.js (hunk 11 above)
Comment (chicharr) [reply src/controllers/suggestions.js:161]: Agreed... will do in a separate PR

File: src/controllers/suggestions.js (hunk 13 above)
Comment (chicharr) [reply src/controllers/suggestions.js:286]: Agreed... will do in a separate PR

File: src/controllers/opportunities.js (hunk 6 above)
Comment (chicharr) [reply src/controllers/opportunities.js]: DONE
We need to decide what to do with data access errors. Do we return them as they come, or do we transform the message somehow?

File: src/controllers/opportunities.js (hunk 5 above)
Comment (chicharr) [reply src/controllers/opportunities.js]: DONE
We need to decide what to do with data access errors. Do we return them as they come, or do we transform the message somehow?

File: src/controllers/opportunities.js (hunk 4 above)
Comment (chicharr) [reply src/controllers/opportunities.js:144]: DONE. 
We need to decide what to do with data access errors. Do we return them as they come, or do we transform the message somehow?

File: src/controllers/opportunities.js (hunk 3 above)
Comment (chicharr) [reply src/controllers/opportunities.js]: it is not needed. Side effects of copy-pasting. 
Removing it

File: src/controllers/opportunities.js (hunk 1 above)
Comment (chicharr) [reply src/controllers/opportunities.js]: DONE

File: src/controllers/suggestions.js (hunk 7 above)
Comment (chicharr) [reply src/controllers/suggestions.js:64]: I'm open to change this, but IMO `badRequest` is the most adequate here, as the problem is trying to get the suggestions of an opportunity that either does not exist or is not bound to the site passed as a parameter.
IMO a 404 would be more appropriate if you try to get a suggestion by ID and the suggestion is not found.

File: src/controllers/suggestions.js (hunk 9 above)
Comment (chicharr) [reply src/controllers/suggestions.js:93]: see above :)

File: src/controllers/suggestions.js (hunk 10 above)
Comment (chicharr) [reply src/controllers/suggestions.js:130]: This API call is to get a suggestion by passing: siteID, opportunityID and suggestionID. IMO if any of the 3 parameters is wrong or there's no relation site -> oppty -> suggestion,  the error should be the same "Suggestion not found"

File: src/controllers/suggestions.js (hunk 8 above)
Comment (chicharr) [reply src/controllers/suggestions.js:62]: Agreed to check first for the Opportunity

File: src/controllers/suggestions.js (hunk 10 above)
Comment (chicharr) [reply src/controllers/suggestions.js]: Agreed to return unspecific "not found" message when the opportunity and the site don't match

File: src/controllers/suggestions.js (hunk 9 above)
Comment (chicharr) [reply src/controllers/suggestions.js:93]: Agreed to check first for the Opportunity.
Will be done in a separate PR

**Review by solaris007 — APPROVED** [review]
LGTM

**Pull request merged**